Modules/_asynciomodule.c

```diff
@@ -3714,10 +3715,11 @@ _asyncio_all_tasks_impl(PyObject *module, PyObject *loop)
         if (add_one_task(state, tasks, (PyObject *)head, loop) < 0) {
             Py_DECREF(tasks);
             Py_DECREF(loop);
+            Py_DECREF(head);
             return NULL;
         }
-        head = head->next;
-        assert(head != NULL);
+        Py_INCREF(head->next);
+        Py_SETREF(head, head->next);
```

Comment:
Just stylistic, so up to you:

```suggestion
        Py_SETREF(head, Py_NewRef(head->next));
```

Reply:
Actually it would require another cast to TaskObj which I am not liking hence leaving it